raft: state.commit is out of range

via local-tester with reordering:

```
2016-06-13 17:47:28.006281 I | etcdmain: etcd Version: 3.0.0-beta.0+git
2016-06-13 17:47:28.006332 I | etcdmain: Git SHA: 65e19a1
2016-06-13 17:47:28.006337 I | etcdmain: Go Version: go1.6
2016-06-13 17:47:28.006348 I | etcdmain: Go OS/Arch: linux/amd64
2016-06-13 17:47:28.006353 I | etcdmain: setting maximum number of CPUs to 8, total number of available CPUs is 8
2016-06-13 17:47:28.006359 W | etcdmain: no data-dir provided, using default data-dir ./infra1.etcd
2016-06-13 17:47:28.006391 N | etcdmain: the server is already initialized as member before, starting as etcd member... 
2016-06-13 17:47:28.006451 I | etcdmain: listening for peers on http://127.0.0.1:12380
2016-06-13 17:47:28.006474 I | etcdmain: listening for client requests on 127.0.0.1:11119 
2016-06-13 17:47:28.013070 I | etcdserver: recovered store from snapshot at index 38626 
2016-06-13 17:47:28.013081 I | etcdserver: name = infra1
2016-06-13 17:47:28.013087 I | etcdserver: data dir = infra1.etcd 
2016-06-13 17:47:28.013093 I | etcdserver: member dir = infra1.etcd/member
2016-06-13 17:47:28.013098 I | etcdserver: heartbeat = 100ms
2016-06-13 17:47:28.013104 I | etcdserver: election = 1000ms
2016-06-13 17:47:28.013109 I | etcdserver: snapshot count = 1000
2016-06-13 17:47:28.013118 I | etcdserver: advertise client URLs = http://127.0.0.1:2379
2016-06-13 17:47:28.065753 I | etcdserver: restarting member 5da0b1f0ade347d1 in cluster ea3db81f3897e3ad at commit index 38040
2016-06-13 17:47:28.065854 C | raft: 5da0b1f0ade347d1 state.commit 38040 is out of range [38626, 38626]
panic: 
5da0b1f0ade347d1 state.commit 38040 is out of range [38626, 38626]
goroutine 1 [running]:
panic(0xcd2fa0, 0xc8205ff170) 
        /usr/lib/go/src/runtime/panic.go:464 +0x3e6
github.com/coreos/etcd/cmd/vendor/github.com/coreos/pkg/capnslog.(*PackageLogger).Panicf(0xc82010d100, 0x119e200, 0x2b, 0xc8204a5d40, 0x4, 0x4)
        /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/pkg/capnslog/pkg_logger.go:73 +0x191
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft.(*raft).loadState(0xc820204340, 0x67a, 0xd2643f51f16cc22b, 0x9498, 0x0, 0x0, 0x0)
        /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft/raft.go:942 +0x2a2
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft.newRaft(0xc8201f7b30, 0x451b20)
        /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft/raft.go:225 +0x8ff
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft.RestartNode(0xc8201f7b30, 0x0, 0x0)
        /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft/node.go:212 +0x45
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver.restartNode(0xc8202163c0, 0xc820208990, 0x29, 0xc8201f7f68, 0x0, 0x0, 0x0, 0x0)
        /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/raft.go:361 +0x7c7
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver.NewServer(0xc8202163c0, 0x0, 0x0, 0x0)
        /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdserver/server.go:350 +0x3cf2
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain.startEtcd(0xc820160800, 0x0, 0x0, 0x0)
        /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain/etcd.go:366 +0x23ea
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain.startEtcdOrProxyV2()
        /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain/etcd.go:116 +0x213d
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain.Main()
        /home/anthony/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/etcdmain/main.go:36 +0x21e
main.main()
        /home/anthony/go/src/github.com/coreos/etcd/cmd/main.go:28 +0x14
Terminating etcd1
```

--- a/etcdserver/raft.go
+++ b/etcdserver/raft.go
@@ -216,18 +216,19 @@ func (r *raftNode) start(s *EtcdServer) {
 					r.s.send(rd.Messages)
 				}
 
+				if err := r.storage.Save(rd.HardState, rd.Entries); err != nil {
+					plog.Fatalf("raft save state and entries error: %v", err)
+				}
+				if !raft.IsEmptyHardState(rd.HardState) {
+					proposalsCommitted.Set(float64(rd.HardState.Commit))
+				}
+
 				if !raft.IsEmptySnap(rd.Snapshot) {
 					if err := r.storage.SaveSnap(rd.Snapshot); err != nil {
 						plog.Fatalf("raft save snapshot error: %v", err)
 					}
 					r.raftStorage.ApplySnapshot(rd.Snapshot)
 					plog.Infof("raft applied incoming snapshot at index %d", rd.Snapshot.Metadata.Index)
-				}
-				if err := r.storage.Save(rd.HardState, rd.Entries); err != nil {
-					plog.Fatalf("raft save state and entries error: %v", err)
-				}
-				if !raft.IsEmptyHardState(rd.HardState) {
-					proposalsCommitted.Set(float64(rd.HardState.Commit))
 				}
 
 				r.raftStorage.Append(rd.Entries)

